src/Processors/QueryPlan/SortingStep.cpp

```diff
@@ -111,129 +114,186 @@ void SortingStep::convertToFinishSorting(SortDescription prefix_description_)
-                    header, prefix_description, result_description, max_block_size, limit, increase_sort_description_compile_attempts_current);
-            });
-        }
-    }
-    else if (type == Type::Full)
+            return std::make_shared<FinishSortingTransform>(
+                header, input_sort_desc, result_sort_desc, max_block_size, limit_, increase_sort_description_compile_attempts_current);
+        });
+}
+
+void SortingStep::mergingSorted(QueryPipelineBuilder & pipeline, const SortDescription & result_sort_desc, const UInt64 limit_)
+{
+    /// If there are several streams, then we merge them into one
+    if (pipeline.getNumStreams() > 1)
     {
-        pipeline.addSimpleTransform([&](const Block & header, QueryPipelineBuilder::StreamType stream_type) -> ProcessorPtr
+        auto transform = std::make_shared<MergingSortedTransform>(
+            pipeline.getHeader(),
+            pipeline.getNumStreams(),
+            result_sort_desc,
+            max_block_size,
+            SortingQueueStrategy::Batch,
+            limit_);
+
+        pipeline.addTransform(std::move(transform));
+    }
+}
+
+void SortingStep::mergeSorting(QueryPipelineBuilder & pipeline, const SortDescription & result_sort_desc, UInt64 limit_)
+{
+    bool increase_sort_description_compile_attempts = true;
+
+    pipeline.addSimpleTransform(
+        [&, increase_sort_description_compile_attempts](
+            const Block & header, QueryPipelineBuilder::StreamType stream_type) mutable -> ProcessorPtr
         {
-            if (stream_type != QueryPipelineBuilder::StreamType::Main)
+            if (stream_type == QueryPipelineBuilder::StreamType::Totals)
                 return nullptr;
 
-            return std::make_shared<PartialSortingTransform>(header, result_description, limit);
+            // For multiple FinishSortingTransform we need to count identical comparators only once per QueryPlan.
+            // To property support min_count_to_compile_sort_description.
+            bool increase_sort_description_compile_attempts_current = increase_sort_description_compile_attempts;
+
+            if (increase_sort_description_compile_attempts)
+                increase_sort_description_compile_attempts = false;
+
+            return std::make_shared<MergeSortingTransform>(
+                header,
+                result_sort_desc,
+                max_block_size,
+                limit_,
+                increase_sort_description_compile_attempts_current,
+                max_bytes_before_remerge / pipeline.getNumStreams(),
+                remerge_lowered_memory_bytes_ratio,
+                max_bytes_before_external_sort,
+                tmp_volume,
+                min_free_disk_space);
         });
+}
+
+void SortingStep::fullSort(QueryPipelineBuilder & pipeline, const SortDescription & result_sort_desc, const UInt64 limit_, const bool skip_partial_sort)
+{
+    if (!skip_partial_sort)
```

Comment:
I'm not sure that it will not introduce regression in some cases. `PartialSortingTransform` can throw out significant portion of rows due to the `limit`. pls, check it carefully (also taking [this](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/Transforms/PartialSortingTransform.h#L44) into account), let's add a perf test for such case.